data-prepper-plugins/obfuscate-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/obfuscation/action/OneWayHashAction.java

```diff
@@ -0,0 +1,118 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@DataPrepperPlugin(name = "hash", pluginType = ObfuscationAction.class, pluginConfigurationType = OneWayHashActionConfig.class)
+public class OneWayHashAction implements ObfuscationAction {
+    
+
+    private MessageDigest messageDigest = null;
+    private byte[] salt;
+    private String saltKey;
+    private static final Logger LOG = LoggerFactory.getLogger(ObfuscationProcessor.class);
+
+    @DataPrepperPluginConstructor
+    public OneWayHashAction(final OneWayHashActionConfig config) {
+        
+        this.salt = config.getSalt();
+        this.saltKey = config.getSaltKey();
+        try{
+            messageDigest = MessageDigest.getInstance(config.getHashFormat());            
+        } catch (NoSuchAlgorithmException noSuchAlgorithmException){
+            LOG.error("There was an exception while processing Event [{}]", noSuchAlgorithmException);
+        }
+    }
+
+    @Override
+    public String obfuscate(String source, List<Pattern> patterns, Record<Event> record) {
+
+        byte [] saltToApply = this.salt;
+                
+        // Resolve salt to compute based on a path provided in the configuration.  
+        // For records where path was not found, the salt value defined in the pipeline configuration will be used, if salt value was not configured, one will be generated.        
+
+        if(saltKey != null && saltKey.equals("") == false){                
+            final Event recordEvent = record.getData();                        
+            if(recordEvent.containsKey(saltKey)){
+                saltToApply = computeSaltBasedOnKeyValue(recordEvent.get(saltKey, String.class));
+            } else {
+                LOG.info("Unable to find a key '{}' for using as salt, using default salt pipeline configuration for the record instead", saltKey);
+            }         
+        }
+
+        if (patterns == null || patterns.size() == 0) {
+            // no pattern to match, replace the whole string
+            return oneWayHashString(source,saltToApply);
+        }
+        
+        String replacementString = source;
+
+        for (Pattern pattern : patterns) {
+            Matcher matcher = Pattern.compile(pattern.pattern()).matcher(replacementString);
+            StringBuffer stringBuffer = new StringBuffer();
+            while (matcher.find()) {
+
+                String stringToHash = source.substring(matcher.start(),matcher.end());
+
+                matcher.appendReplacement(stringBuffer, oneWayHashString(stringToHash,saltToApply));
+            }
+            matcher.appendTail(stringBuffer);
+            replacementString = stringBuffer.toString();
+        }
+        return replacementString;
+        
+        
+    }
+
+    private String oneWayHashString(String source, byte[] salt){
+        String oneWayHashedSource = "";
+
+        try {
+            MessageDigest messageDigestClone = (MessageDigest) messageDigest.clone();
```

Comment:
You only need to clone this when the salt is provided via the `salt_key`. For other approaches, you can re-use the `MessageDigest`. Also, you don't need to re-clone when using this in the loop for `patterns` (line 77 above). This should help with performance in those cases.

Also, you could probably help even the `salt_key` cases by using a `ThreadLocal` `MessageDigest`. This might require some additional effort, so may we leave it for a later optimization.

Reply:
@dlvenable Please remind me what will happen when you have 2+ hash processors in the pipeline. One will hash field ABC with salt S1 and another will hash field XYZ with salt S2. Will those plugin instances initialize individually where I can get distinct messageDigest instances?

example pipeline:
```
pipeline:
  source:
...
  processor:
    - parse_json:
    - obfuscate:
        source: "log"
        target: "new_log"
        patterns:
          - "[A-Za-z0-9+_.-]+@([\\w-]+\\.)+[\\w-]{2,4}"
        action:
          hash:
            salt: "AAAAAAAAAAAAAAAA"
    - obfuscate:
        source: "phone"
        action:
          hash:
            salt: "BBBBBBBBBBBBBBBB"
  sink:
    - stdout:
```

Reply:
@mishavay-aws , For the scenario you outlined with two different `obfuscate` processors, Data Prepper will create two instances of `OneWayHashAction`. There will be no shared objects between the two.